### ISSUE TITLE
click+ctrl goes to all references/definition

### ISSUE THREAD

**Issue by SymphonyNineth** [zed-industries/zed#12752]
### Check for existing issues

- [x] Completed

### Describe the bug / provide steps to reproduce it

Clicking on entity and afterwards pressing ctrl goes finds all the references(as if clicking while ctrl is pressed) instead of just not doing it :). I assume is not the expected behavior since it makes you wait for a while before you can use ctrl key after clicking somewhere and i constantly fall into the references tab.

### Environment

Zed: v1.0.0 (Zed Dev 48581167b73165fe9f9c780a73a6be7c1602f7a9)
OS: Linux 1.0.0
Memory: 31.1 GiB
Architecture: x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

For example here i click on SingleLine and then press ctrl to be able to delete the entire word right away with ctrl + backspace.
![image](https://github.com/zed-industries/zed/assets/63275027/f04d7902-985f-43d6-8c5d-8bb9fbd90170)
I expect it to not go to the references because the ctrl key was not pressed at the time of clicking, but that's what the editor does.
![image](https://github.com/zed-industries/zed/assets/63275027/b4813fcc-c3ce-4da7-89ba-214ac39a24ea)

**Comment by Aeledfyr** [comment]
The issue title is easy to misinterpret—the specific issue is that pressing control *after* clicking still triggers go to definition/references.

I believe the root cause of this issue is that following links (go to def) runs on mouse up if control is held, *even if control was not held at mouse down*.  As such, it is easy to accidentally trigger go to definition by pressing control immediately after clicking.

From experimenting a bit, it appears the VSCode handles this by only following the link if control was held at both mouse down and mouse up (though it doesn't check in-between).

Specifically, this is in `EditorElement::mouse_up`:
https://github.com/zed-industries/zed/blob/d5a6ca4914b5f009223df99a57cd799f2af4b470/crates/editor/src/element.rs#L578-L588

(Note that `multi_cursor_modifier` is a misnomer -- it's actually whichever modifier *isn't* the multi-cursor modifier)

It should be relatively straightforward to track whether the modifier key was held at the start of the mouse press and only trigger the action if the modifier was held at both the start and end of the press.

**Comment by SymphonyNineth** [comment]
> The issue title is easy to misinterpret—the specific issue is that pressing control _after_ clicking still triggers go to definition/references.
> 
> I believe the root cause of this issue is that following links (go to def) runs on mouse up if control is held, _even if control was not held at mouse down_. As such, it is easy to accidentally trigger go to definition by pressing control immediately after clicking.
> 
> Specifically, this is in `EditorElement::mouse_up`:
> 
> https://github.com/zed-industries/zed/blob/d5a6ca4914b5f009223df99a57cd799f2af4b470/crates/editor/src/element.rs#L578-L588
> 
> (Note that `multi_cursor_modifier` is a misnomer -- it's actually whichever modifier _isn't_ the multi-cursor modifier)
> 
> From experimenting a bit, it appears the VSCode handles this by only following the link if control was held at both mouse down and mouse up (though it doesn't check if it was held between those points).
> 
> It should be relatively straightforward to track whether the modifier key was held at the start of the mouse press and only trigger the action if the modifier was held at both the start and end of the press.

That could very well be the issue, however sadly i cannot patch it myself, i'm not very familiar with Rust(

**Comment by azomDev** [comment]
I opened another issue with (in my opinion) a more accurate title. I closed it since I found this issue later and it is the same problem.

If there is anything I can do to help, let me know. Later when I'll have more time i'll look to see if it's an issue I can try fixing, but for now assume I can't.

**Comment by smoores-dev** [comment]
Hi! @JosephTLyons I noticed that this was just added to the GitHub Project — I'd like to take a stab at it if no one else is already planning to work on it! I started messing around with it locally a few hours ago and I _think_ I have an idea for a reasonable approach. Let me know if you're interested in accepting a PR for this (and/or if anyone has a moment to chat with me about what I'm thinking)!

**Comment by JosephTLyons** [comment]
@smoores-dev, absolutely. We haven't started in on a quality week and even if we did, as long as we try avoid situations where multiple people are accidentally working on the same thing, then everything is fair game. We'd take a PR on this. As for asking for advice / spitballing some ideas, you'll likely have to chat with one of the more dedicated Zed engineers, like @mikayla-maki, @SomeoneToIgnore, or @osiewicz.

**Comment by smoores-dev** [comment]
Awesome, thanks for the info! I think I'll have at least a draft to review by next week. Would it be best if I started a discussion to chat about the approach and tagged some of the engineers?

**Comment by smoores-dev** [comment]
It seems that this is back 😩. I have not yet reasoned through how. I saw that there was some refactoring in #35075 to the relevant code, but it doesn't seem like that's actually been released yet. I'm on 0.199.10, and if I mousedown, then press and hold ctrl, then mouseup, I trigger "go to definition" as if I had held ctrl during the mousedown.

**Comment by smoores-dev** [comment]
@JosephTLyons I just wanted to make sure you saw this, it seems like a recent release (I'm not sure which one yet) reintroduced this bug!

**Comment by jeppester** [comment]
I face this issue all the time on my laptop. I think it might be worse with touchpads.

I get it very often when I:
- Join lines
- Select next/all
- Select a line I want to copy
- Duplicate lines (I have ctrl + shift + d set up as duplicate down)

It's especially annoying when the ctrl-command is altering the code (like joining lines). The command is then triggered in the definition file, forcing you to revert the change, or the definition file cannot be closed easily.

**Comment by smoores-dev** [comment]
I have something of a long queue in terms of open source updates to get to, but I will try to dedicate some time soon to figuring out how this regressed and seeing if I can fix it. It is indeed _very_ annoying, especially with trackpads!

**Comment by jjf21** [comment]
Hi, shouldn't this ticket be re-opened ? This bug is really annoying

**Comment by smoores-dev** [comment]
I opened a PR that changes the behavior back to that of my original PR. Hopefully this time with sufficient context to prevent it from being reverted again 😬.

If this is really driving folks nuts, you can make a local build from [my branch](https://github.com/smoores-dev/zed/tree/click-event-modifiers-2) — I'll try to keep it up-to-date at least weekly.
